File: src/SlimShader/Shader/Tokens/CustomDataToken.cs
using System;
using SlimShader.IO;
using SlimShader.Util;

namespace SlimShader.Shader.Tokens
{
	/// <summary>
	/// Custom-Data Block Format
	///
	/// DWORD 0 (CustomDataDescTok):
	/// [10:00] == D3D10_SB_OPCODE_CUSTOMDATA
	/// [31:11] == D3D10_SB_CUSTOMDATA_CLASS
	///
	/// DWORD 1: 
	///          32-bit unsigned integer count of number
	///          of DWORDs in custom-data block,
	///          including DWORD 0 and DWORD 1.
	///          So the minimum value is 0x00000002,
	///          meaning empty custom-data.
	///
	/// Layout of custom-data contents, for the various meta-data classes,
	/// not defined in this file.
	/// </summary>
	public abstract class CustomDataToken : OpcodeToken
	{
		public CustomDataClass CustomDataClass { get; private set; }

		public static CustomDataToken Parse(BytecodeReader reader, uint token0)
		{
			var customDataClass = token0.DecodeValue<CustomDataClass>(11, 31);
			CustomDataToken token;
			switch (customDataClass)
			{
				case CustomDataClass.DclImmediateConstantBuffer:
					token = ImmediateConstantBufferDeclarationToken.Parse(reader);
					break;
				case CustomDataClass.ShaderMessage:
					token = ShaderMessageDeclarationToken.Parse(reader);
					break;
				default:
					throw new ArgumentOutOfRangeException();
			}

			token.CustomDataClass = customDataClass;
			return token;
		}
	}
}

File: src/cpp/Source/Chunks/Shex/Tokens/CustomDataToken.cpp
#include "PCH.h"
#include "CustomDataToken.h"

#include "Decoder.h"
#include "ImmediateConstantBufferDeclarationToken.h"
#include "ShaderMessageDeclarationToken.h"

using namespace std;
using namespace SlimShader;

shared_ptr<CustomDataToken> CustomDataToken::Parse(BytecodeReader& reader, uint32_t token0)
{
	auto customDataClass = DecodeValue<CustomDataClass>(token0, 11, 31);
	shared_ptr<CustomDataToken> token;
	switch (customDataClass)
	{
	case CustomDataClass::DclImmediateConstantBuffer:
		token = ImmediateConstantBufferDeclarationToken::Parse(reader);
		break;
	case CustomDataClass::ShaderMessage:
		token = ShaderMessageDeclarationToken::Parse(reader);
		break;
	default:
		throw runtime_error("Unrecognised custom data class");
	}

	token->_customDataClass = customDataClass;
	return token;
}